### PR TITLE
Expand `$ref` when assigning the `definition` field to `JsonSchemaValueException`

### LINKED ISSUE #141: Exceptions are obscure when using `$ref` inside `anyOf`, `oneOf`, `allOf`, etc...
Hello, when validating composite schemas with `fastjsonschema` (i.g. `oneOf`, `allOf`, etc..) the error messages can be very tricky to understand. In those scenarios I always try to reconstruct some meaningful error messages by looking at `exception_obj.definition`.

However, that is very hard to do when using `$ref` to organise the schemas in smaller pieces, specially if those smaller pieces are stored in separated schemas that get re-used.
This happens because both the error message and the exception attribute will only briefly mention the value of `$ref` itself, instead of the actual schema that is not being matched...

This behaviour is shown in the example bellow:

```pycon
>>> from pprint import pprint
>>>
>>> import fastjsonschema
>>>
>>> repo = {
...     "sch://m3": {"type": "number", "multipleOf": 3},
...     "sch://m5": {"type": "number", "multipleOf": 5},
... }
>>> schema = {
...     "$schema": "http://json-schema.org/draft-07/schema",
...     "anyOf": [
...         {"$ref": "sch://m3"},
...         {"$ref": "sch://m5"},
...     ]
... }
>>>
>>> fastjsonschema.validate(schema, 2, handlers={"sch": repo.__getitem__})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "~/workspace/python-fastjsonschema/fastjsonschema/__init__.py", line 114, in validate
    return compile(definition, handlers, formats, use_default)(data)
  File "<string>", line 16, in validate
fastjsonschema.exceptions.JsonSchemaValueException: data must be valid by one of anyOf definition
>>>
>>> try:
...    fastjsonschema.validate(schema, 2, handlers={"sch": repo.__getitem__})
... except fastjsonschema.JsonSchemaValueException as ex:
...     print(ex)
...     print("Rule:", ex.rule)
...     print("Definition:")
...     pprint(ex.definition)
...
data must be valid by one of anyOf definition
Rule: anyOf
Definition:
{'$schema': 'http://json-schema.org/draft-07/schema',
 'anyOf': [{'$ref': 'sch://m3'}, {'$ref': 'sch://m5'}]}
>>>
```

Please note that the user seeing the error, or the API consumer needs to have access to all the JSONSchema documents to understand what is going wrong...

We can compare this behaviour with the other popular library:

```pycon
>>> import jsonschema
>>> resolver = jsonschema.RefResolver.from_schema({}, handlers={"sch": repo.__getitem__})
>>> jsonschema.validate(2, schema=schema, resolver=resolver)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "~/workspace/python-fastjsonschema/.venv/lib/python3.8/site-packages/jsonschema/validators.py", line 1022, in validate
    raise error
jsonschema.exceptions.ValidationError: 2 is not a multiple of 3

Failed validating 'multipleOf' in schema[0]:
    {'multipleOf': 3, 'type': 'number'}

On instance:
    2
>>>
```

Although the error here is more imprecise than the one provided by `fastjsonschema`, it is also more clear because it prints the offending rule, expanding `$ref`.

### TLDR
If `fastjsonschema` expand `$ref` in the `exception_obj.definitions`, it would be much easier to understand, deal with value exceptions...

### PR DESCRIPTION
**Pull request by abravalheri**
This change should fix #141.

By expanding any `$ref` field in the schema definition that is available in the `JsonSchemaValueException` objects, we make it easier for the API consumers to provide more meaningful errors.

### CODE REVIEW

File: fastjsonschema/generator.py [hunk 1]
```diff
@@ -245,8 +245,27 @@ def exc(self, msg, *args, rule=None):
...
+        if isinstance(definition, list):
+            return [self._expand_refs(v) for v in definition]
+        if not isinstance(definition, dict):
+            return definition
+        if "$ref" in definition:
+            with self._resolver.resolving(definition["$ref"]) as schema:
+                return schema
+        if "properties" in definition:
+            return {
+                **definition,
```
Comment (horejsek) [inline fastjsonschema/generator.py:262]: Shouldn't definition here also recursively expand refs?

**Comment by abravalheri** [comment]
Thank you very much for the kind review @horejsek.

I don't know about [this question](https://github.com/horejsek/python-fastjsonschema/pull/142#discussion_r814725816), but I changed the code anyway as per your suggestion... I just had to add an additional condition to avoid failures, but then the tests pass.

**Comment by abravalheri** [comment]
Question you be interested in another PR with improved error messages?

File: fastjsonschema/generator.py (hunk 1 above)
Comment (horejsek) [reply fastjsonschema/generator.py]: Tried to look around and found scenario here: https://json-schema.org/understanding-json-schema/reference/object.html#unevaluated-properties

```
{
  "type": "object",
  "properties": {
    "street_address": { "type": "string" },
    "city": { "type": "string" },
    "state": { "type": "string" },
    "type": { "enum": ["residential", "business"] }
  },
  "required": ["street_address", "city", "state", "type"],

  "if": {
    "type": "object",
    "properties": {
      "type": { "const": "business" }
    },
    "required": ["type"]
  },
  "then": {
    "properties": {
      "department": { "type": "string" }
    }
  },

  "unevaluatedProperties": false
}
```

In other words, if there are properties, there could be also if/then which can also contain refs. So I think that deserves tests on its own and do expansion here as well. Do you agree or maybe I have it wrong and this is enough? Maybe in either case test like that would be helpful so we can see its not needed as well. Do you think you can do that?

Comment (abravalheri) [reply fastjsonschema/generator.py]: Yeah, I agree that it is a good idea to add a test case for that.

Let me add it ;)

Comment (horejsek) [reply fastjsonschema/generator.py]: Thank you, appreciated!

**Comment by horejsek** [comment]
Sure, error messages need some improvements. There are also some open issues for it, so might be useful looking there as well. If you are willing to improve it, let me know what you have in mind if you want to discuss it first, and feel free to open PR for it.

**Comment by abravalheri** [comment]
Thank you very much @horejsek, I added [tests](https://github.com/horejsek/python-fastjsonschema/pull/142/commits/64ac97e9d2d19154c4afc0836de9005a90140a23) using `$ref` inside `if-then-else`.


(I also added a regular [example](https://github.com/horejsek/python-fastjsonschema/pull/142/commits/cd13959062fd05ec79433baa4fcd75fd8afc0599) with `if-then-else`, just as a sanity check before working with `$ref` because I was having some problems with the syntax)

**Comment by abravalheri** [comment]
I will have a look on the open issues and try to do some error message improvements in the future. Thank you very much!

**Pull request merged**